File: sources/CSharp/src/Ers/Visualization/Window.cs
using System.Runtime.InteropServices;
using Ers.Engine;

namespace Ers
{
    public class Window
    {
        private IntPtr coreInstance;

        public Window(IntPtr windowHandle, IntPtr displayHandle, int width, int height)
        {
            this.coreInstance = ErsEngine.ERS_Window_Create(windowHandle, displayHandle, width, height);
        }

        public IntPtr GetCoreInstance() { return coreInstance; }

        public void DrawRenderContext(RenderContext renderContext)
        {
            ErsEngine.ERS_Window_DrawRenderContext(coreInstance, renderContext.GetCoreInstance());
        }

        public void Present() { ErsEngine.ERS_Window_Present(coreInstance); }

        public void DisposeInner()
        {
            if (coreInstance != IntPtr.Zero)
            {
                ErsEngine.ERS_Window_Destroy(coreInstance);
                coreInstance = IntPtr.Zero;
            }
        }

        ~Window() => DisposeInner();
    }
}

File: sources/CPP/include/Ers/Visualization/Window.cpp
#include "Window.h"
#include "Ers/Api.h"

namespace Ers
{
    Window::Window(void* windowHandle, void* displayHandle, int screenWidth, int screenHeight)
    {
        data = ersAPIFunctionPointers.ERS_Window_Create(windowHandle, displayHandle, screenWidth, screenHeight);
    }

    void Window::Present()
    {
        ersAPIFunctionPointers.ERS_Window_Present(data);
    }

    void Window::DrawRenderContext(RenderContext& renderContext)
    {
        ersAPIFunctionPointers.ERS_Window_DrawRenderContext(data, renderContext.Data());
    }

    Window::~Window()
    {
        ersAPIFunctionPointers.ERS_Window_Destroy(data);
    }

} // namespace Ers
